File: 10871 Primed Subsequence/solution.java
import java.io.*;
import java.util.*;

public class Main {
    static int MAX_NUM = 10000;
    static boolean[] isComposite = new boolean[MAX_NUM + 1];
    static ArrayList<Short> primes = new ArrayList<>();
    public static void main(String[] args) throws IOException {
        eulerSieve();
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StreamTokenizer st = new StreamTokenizer(br);
        st.nextToken();
        int t = (int)st.nval;
        StringBuilder output = new StringBuilder();
        while(t-- > 0) {
            st.nextToken();
            int n = (int)st.nval;
            int[][] numbers = new int[2][n];
            for(int i = 0; i < n; ++i) {
                st.nextToken();
                numbers[0][i] = numbers[1][i] = (int)st.nval;
            }
            int length = 2;
            boolean antiPrimed = true;
            while(antiPrimed && length <= n) {
                for(int i = 0; i <= n - length && antiPrimed; ++i) {
                    numbers[1][i] += numbers[0][i + length - 1];
                    if((numbers[1][i] <= MAX_NUM && !isComposite[numbers[1][i]]) || (numbers[1][i] > MAX_NUM && isPrime(numbers[1][i]))) {
                        output.append("Shortest primed subsequence is length ").append(length).append(':');
                        for(int j = i; j < i + length; ++j)
                            output.append(' ').append(numbers[0][j]);
                        output.append('\n');
                        antiPrimed = false;
                    } 
                }
                ++length;
            }
            if(antiPrimed)
                output.append("This sequence is anti-primed.\n");
        }
        System.out.print(output);
    }
    
    static void eulerSieve() {
	    isComposite[0] = isComposite[1] = true;
        for(short n = 2; n <= MAX_NUM; ++n) {
            if(!isComposite[n])
                primes.add(n);
            for(int i = 0, temp; (temp = primes.get(i) * n) <= MAX_NUM; ++i) {
                isComposite[temp] = true;
                if(n % primes.get(i) == 0)
                    break;
            }
        }
    }
    
    static boolean isPrime(int n) {
        int limit = (int)Math.sqrt(n);
        for(int i = 0; i < primes.size() && primes.get(i) <= limit; ++i) {
            if(n % primes.get(i) == 0)
                return false;
        }
        return true;
    }
}

File: 10871 Primed Subsequence/solution.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <stdbool.h>
#define MAX_NUM 10000
#define GET(n) (mark[(n) >> 5] & (1u << ((n) & 31)))
#define SET(n) (mark[(n) >> 5] |= (1u << ((n) & 31)))

unsigned int mark[(MAX_NUM >> 5) + 1];
short *primes;
int size;

void eulerSieve() {
    SET(0);
    SET(1);
    for(short n = 2; n <= MAX_NUM; ++n) {
        if(!GET(n)) {
            primes = (short*)realloc(primes, (size + 1) * sizeof(short));
            primes[size++] = n;
        }
        for(int i = 0, temp; (temp = primes[i] * n) <= MAX_NUM; ++i) {
            SET(temp);
            if(n % primes[i] == 0)
                break;
        }
    }
}

bool isPrime(int n) {
    int limit = sqrt(n);
    for(int i = 0; i < size && primes[i] <= limit; ++i) {
        if(n % primes[i] == 0)
            return false;
    }
    return true;
}

int main() {
    eulerSieve();
    int t;
    scanf("%d", &t);
    while(t--) {
        int n;
        scanf("%d", &n);
        int numbers[2][n];
        for(int i = 0; i < n; ++i) {
            scanf("%d", &numbers[0][i]);
            numbers[1][i] = numbers[0][i];
        }
        int length = 2;
        bool antiPrimed = true;
        while(antiPrimed && length <= n) {
            for(int i = 0; i <= n - length && antiPrimed; ++i) {
                numbers[1][i] += numbers[0][i + length - 1];
                if((numbers[1][i] <= MAX_NUM && !GET(numbers[1][i])) || (numbers[1][i] > MAX_NUM && isPrime(numbers[1][i]))) {
                    printf("Shortest primed subsequence is length %d:", length);
                    for(int j = i; j < i + length; ++j)
                        printf(" %d", numbers[0][j]);
                    putchar('\n');
                    antiPrimed = false;
                } 
            }
            ++length;
        }
        if(antiPrimed)
            puts("This sequence is anti-primed.");
    }
    free(primes);
    return 0;
}
